autrainer/core/scripts/preprocess_script.py

```diff
@@ -196,39 +197,96 @@ def _process_chunk(
         ):
             print(f" - {name}")
-            if preprocess is None or os.path.isdir(
-                os.path.join(dataset["path"], dataset["features_subdir"])
-            ):
+            # swap dataset handler with preprocessing handler
+            features_subdir = dataset["features_subdir"]
+            output_file_handler = autrainer.instantiate_shorthand(
+                config=dataset["file_handler"],
+                instance_of=AbstractFileHandler,
+            )
+            output_file_type = dataset["file_type"]
+            # set features_subdir to None (defaults to audio)
+            # so that iteration takes place over raw data
+            dataset["features_subdir"] = None
+            dataset["file_handler"] = preprocess["file_handler"]
+            dataset["file_type"] = preprocess["file_type"]
+            if preprocess is None:
                 continue
-
-            csvs = _get_csvs(dataset["path"])
-            dfs = _read_csvs(dataset["path"], csvs)
-            unique_files = _get_unique_files(dfs, dataset["index_column"])
-            _create_subdirs(
-                os.path.join(dataset["path"], dataset["features_subdir"]),
-                unique_files,
+            data = autrainer.instantiate_shorthand(dataset)
+            loader = torch.utils.data.DataLoader(
+                torch.utils.data.ConcatDataset(
+                    data.train_dataset,
+                    data.dev_dataset,
+                    data.test_dataset
+                ),
+                shuffle=False,
+                num_workers=self.num_workers,
+                batch_size=1  #TODO: can we do it batched?
+            )
+            pipeline = SmartCompose(
+                [
+                    autrainer.instantiate_shorthand(t)
+                    for t in preprocess["pipeline"]
+                ]
             )
-            chunks = _split_chunks(unique_files, self.num_workers)
-            lock = tqdm.get_lock()
-            with tqdm(
-                total=len(unique_files),
-                desc=name,
-                disable=self.update_frequency == 0,
-            ) as pbar:
-                with ThreadPoolExecutor(self.num_workers) as executor:
-                    futures = [
-                        executor.submit(
-                            _process_chunk,
-                            chunk,
-                            dataset,
-                            preprocess,
-                            pbar,
-                            lock,
-                        )
-                        for chunk in chunks
-                    ]
-                    for future in futures:
-                        future.result()
+            for d, n in (
+                (data.train_dataset, "train"),
+                (data.dev_dataset, "dev"),
+                (data.test_dataset, "test"),
+            ):
+                loader = torch.utils.data.DataLoader(
```

Comment:
Yes, we can do it batched in theory. We can get the current batch size directly from the main configuration file as well.

Reply:
Do all transforms work in batched mode though?

Reply:
No our transforms only support a single feature, e.g. we would need to set our preprocessing pipeline during dataset instantiation for `train_transform`, `dev_transform`, and `test_transform`. Then the loader would already receive processed features ready to be stored (however this needs to account for the inference batch size then).

Reply:
Yes, I would avoid using the `train_transform` in particular, since it will most likely include a `RandomCrop`

Reply:
This referred to using the actual preprocessing transform as the `train_transform` etc. (ignoring the ones in the dataset configuration). This way we dont have to manually apply the transform in the preprocessing script, can use the batch size in the main configuration, and potentially not require modification later on when we switch to a dataclass for feature management here.

Reply:
Got it, yes that makes sense! I'm doing an initial refactoring now, let's see how that goes and maybe we adopt this afterwards

Reply:
Okay, you can have a look at def0dda which should solve several of these issues by adding an `audio_subdir` parameter, which is used as `features_subdir` if that is set to `None`, and also solves the dataset instantiation issues you mentioned.

Reply:
i will close this as there is a comment for the new changes.